sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -81,72 +208,52 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
     Matrices Using K-Simplex Numbers" by Andrew Nystrom and John Hughes.
     """
 
-    assert degree in (2, 3)
-
-    if degree == 2:
-        expanded_dimensionality = int((d**2 + d) / 2 - interaction_only*d)
-    else:
-        expanded_dimensionality = int((d**3 + 3*d**2 + 2*d) / 6
-                                      - interaction_only*d**2)
-    if expanded_dimensionality == 0:
-        return None
-    assert expanded_dimensionality > 0
+    # Make the arrays that will form the CSR matrix of the expansion.
+    cdef INDEX_A_t row_i, row_starts, row_ends, i, j, k, i_ptr, j_ptr, k_ptr
 
-    cdef INDEX_T total_nnz = 0, row_i, nnz
+    cdef INDEX_B_t expanded_index=0, num_cols_in_row, col
 
-    # Count how many nonzero elements the expanded matrix will contain.
+    result_indptr[0] = indptr[0]
     for row_i in range(indptr.shape[0]-1):
-        # nnz is the number of nonzero elements in this row.
-        nnz = indptr[row_i + 1] - indptr[row_i]
-        if degree == 2:
-            total_nnz += (nnz ** 2 + nnz) / 2 - interaction_only * nnz
-        else:
-            total_nnz += ((nnz ** 3 + 3 * nnz ** 2 + 2 * nnz) / 6
-                          - interaction_only * nnz ** 2)
-
-    # Make the arrays that will form the CSR matrix of the expansion.
-    cdef cnp.ndarray[DATA_T, ndim=1] expanded_data = cnp.ndarray(
-        shape=total_nnz, dtype=data.dtype)
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indices = cnp.ndarray(
-        shape=total_nnz, dtype=indices.dtype)
-    cdef INDEX_T num_rows = indptr.shape[0] - 1
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indptr = cnp.ndarray(
-        shape=num_rows + 1, dtype=indptr.dtype)
-
-    cdef INDEX_T expanded_index = 0, row_starts, row_ends, i, j, k, \
-                 i_ptr, j_ptr, k_ptr, num_cols_in_row,  \
-                 expanded_column
-
-    with nogil:
-        expanded_indptr[0] = indptr[0]
-        for row_i in range(indptr.shape[0]-1):
-            row_starts = indptr[row_i]
-            row_ends = indptr[row_i + 1]
-            num_cols_in_row = 0
-            for i_ptr in range(row_starts, row_ends):
-                i = indices[i_ptr]
-                for j_ptr in range(i_ptr + interaction_only, row_ends):
-                    j = indices[j_ptr]
-                    if degree == 2:
-                        col = _deg2_column(d, i, j, interaction_only)
-                        expanded_indices[expanded_index] = col
-                        expanded_data[expanded_index] = (
-                            data[i_ptr] * data[j_ptr])
+        row_starts = indptr[row_i]
```

Comment:
I think it's important to keep the `with nogil` here.

Reply:
Good catch